zuuler/ryhter.py

```diff
@@ -0,0 +1,160 @@
+    repositories = [package.get('upstream') for package in packages]
+
+    release = znoyder.get_releases(tag='osp-17.0')[0].get('upstream_release')
+    branch = f'stable/{release}'
+
+    templates_repository = 'https://opendev.org/openstack/openstack-zuul-jobs'
+    templates_branch = 'master'
+    templates_urls = descarger.download_zuul_config(
+        repository=templates_repository,
+        branch=templates_branch,
+        destination=UPSTREAM_CONFIGS_DIR,
+        errors_fatal=False,
+        ignore_existing=True
+    )
+    templates_directory = list(templates_urls.keys())[0]
+
+    directories = []
+    for repository in repositories:
+        project_urls = descarger.download_zuul_config(
+            repository=repository,
+            branch=branch,
+            destination=UPSTREAM_CONFIGS_DIR,
+            errors_fatal=False,
+            ignore_existing=True
+        )
+        for directory in project_urls.keys():
+            directories.append(directory)
+
+    return [templates_directory] + directories
+
+
+def list_existing_osp_projects() -> list:
+    templates_directory = 'openstack/openstack-zuul-jobs'
+
+    directories = [os.path.join(namespace, project)
+                   for namespace in os.listdir(UPSTREAM_CONFIGS_DIR)
+                   for project in os.listdir(os.path.join(UPSTREAM_CONFIGS_DIR,
+                                                          namespace))]
+    if templates_directory in directories:
+        directories.remove(templates_directory)
+
+    return [templates_directory] + directories
+
+
+def process_arguments() -> Namespace:
+    parser = ArgumentParser()
+    parser.add_argument(
+        '-e', '--existing',
+        dest='existing',
+        default=True,
+        action='store_false',
+        help='use existing configs to generate jobs files (default)'
+    )
+    parser.add_argument(
+        '-d', '--download',
+        dest='download',
+        default=False,
+        action='store_true',
+        help='download the zuul configuration files from repositories'
+    )
+    parser.add_argument(
+        '-g', '--generate',
+        dest='generate',
+        default=False,
+        action='store_true',
+        help='generate new zuul configuration files from upstream sources'
+    )
+    parser.add_argument(
+        '-a', '--all', '--collect-all',
+        dest='collect_all',
+        default=False,
+        action='store_true',
+        help='collect all jobs when generating downstream configuration'
+    )
+
+    arguments = parser.parse_args()
+    return arguments
+
+
+def main() -> None:
+    args = process_arguments()
+
+    if args.download or not(os.path.exists(UPSTREAM_CONFIGS_DIR)):
+        print('Downloading new Zuul configuration from upstream...')
+        directories = fetch_osp_projects()
+        templates_directory = directories.pop(0)
+    else:
+        print('Using local Zuul configuration files...')
```

Comment:
Do we want to mention what path is used by default for "local zuul configuration files"?

Reply:
Probably that path finally would be a required parameter to the tool, but I think you are right, printing that would be nice – I created TODO item, thanks.